Fixes #8551 - host recieves correct smart_class_parameter value when multiple overrides in inherited Host Groups exist for parameter

orrabin opened the pull request.

orrabin committed:
Fixes #8551 - host recieves correct smart_class_parameter value when multiple overrides in inherited Host Groups exist for parameter

orrabin commented on app/services/classification/base.rb:
```diff
@@ -36,7 +36,7 @@ def possible_value_orders
     def values_hash(options = {})
       values = Hash.new { |h,k| h[k] = {} }
-      all_lookup_values = LookupValue.where(:match => path2matches).where(:lookup_key_id => class_parameters)
+      all_lookup_values = LookupValue.where(:match => path2matches).order('match DESC').where(:lookup_key_id => class_parameters)
```
`path2matches` returned the matchers in the correct order (the more specific hostgroup match came first) but the `where` returned the order of lookup_values by id
 if the less specific matcher was created first it would be first in `all_lookup_values` and the `sort_by` in line 42 would keep it first 
after that in `update_generic_matcher` it was picked as the highest priority matcher and the host would get the override value of the less specific hostgroup

orrabin commented on app/services/classification/base.rb:
```diff
@@ -36,7 +36,7 @@ def possible_value_orders
+      all_lookup_values = LookupValue.where(:match => path2matches).order('match DESC').where(:lookup_key_id => class_parameters)
       class_parameters.each do |key|
         lookup_values_for_key = all_lookup_values.where(:lookup_key_id => key.id, :use_puppet_default => false)
         sorted_lookup_values = lookup_values_for_key.sort_by { |lv| key.path.index(lv.match.split(LookupKey::EQ_DELM).first) }
```
an alternative to the `order('match DESC')` in line 39 would be to change this to:

``` ruby
sorted_lookup_values = lookup_values_for_key.sort_by { |lv| [key.path.index(lv.match.split(LookupKey::EQ_DELM).first), -1 * lv.match.split(LookupKey::EQ_DELM).last.length] }
```

since the more specific hostgroup has a longer title it would also work

orrabin commented:
unfortunately I found two solutions that both rely on the fact that inherited hostgroups have longer titles than their parent
I'll add a test now but I wanted to push this now in case anyone has a better idea (or a preference between the two) 
background - the more specific hostgroup is found correctly by using `host.send` with hostgroup but is lost in the sql that gets all lookup_values

domcleal commented:
Sorry, I've not tried to fully understand this code yet, but my immediate reaction is that this fix only works for the case of host groups and nested host groups.  I noticed that the problem occurs even with a hostgroup= and fqdn= matcher (http://projects.theforeman.org/issues/8551#note-5), so relying on length-based ordering just isn't going to fix it.  It looks like a deeper problem?

orrabin commented:
@domcleal I tried adding a matcher for a different fqdn and my host didn't get that value, it got the correct one.
Can you please try your example on my code? or send me the exact matchers and values so I can try?
I followed the steps you sent and added a matcher fqdn=`another_host`, value = `[{"name":"VCENTER","value":"fqdn"}].`
I also used `merge_duplicates` and got the two values from hostgroup but not the value I added for fqdn

domcleal commented on app/services/classification/base.rb:
```diff
@@ -36,7 +36,7 @@ def possible_value_orders
+      all_lookup_values = LookupValue.where(:match => path2matches).order('lookup_values.match DESC').where(:lookup_key_id => class_parameters)
       class_parameters.each do |key|
         lookup_values_for_key = all_lookup_values.where(:lookup_key_id => key.id, :use_puppet_default => false)
         sorted_lookup_values = lookup_values_for_key.sort_by { |lv| key.path.index(lv.match.split(LookupKey::EQ_DELM).first) }
```
I'm a little worried about using a .order on line 39 on key=value (match), followed by a .sort_by on keys, in case the sort doesn't retain the sorting of the values.  (It appears to, but I don't think it's defined that it must.)

Should we remove the order() and extend the sort_by block to sort by both the path and the value (when the path keys are identical)?

orrabin replied:
@domcleal when I pushed the test my comments where removed but you answered my question if I should use order or change the sort_by. I'll change this later today.

domcleal replied:
Ah yes, I see the "outdated" comments now.  Right :)

domcleal commented:
Sorry about that, I can't replicate it today and after going through the code in detail, I can't see how it could have happened.  Please ignore the noise!

Works great and makes sense now.  Just one comment above about the two sorting methods, and I think it's ready.

orrabin commented on app/services/classification/base.rb:
```diff
@@ -39,7 +39,7 @@ def values_hash(options = {})
       class_parameters.each do |key|
         lookup_values_for_key = all_lookup_values.where(:lookup_key_id => key.id, :use_puppet_default => false)
-        sorted_lookup_values = lookup_values_for_key.sort_by { |lv| key.path.index(lv.match.split(LookupKey::EQ_DELM).first) }
+        sorted_lookup_values = lookup_values_for_key.sort_by { |lv| [key.path.index(lv.match.split(LookupKey::EQ_DELM).first), -1 * lv.match.split(LookupKey::EQ_DELM).last.length] }
```
@domcleal thanks for the testing and the comments.
I changed the `sort_by` but there might be a better way then using the `-1`.
I wanted to order by index and then by descending length in one `sort_by`

ohadlevy replied:
each with index is a bit slower, have a look at
https://github.com/JuanitoFatas/fast-ruby

domcleal replied:
@orrabin perhaps just reformat this so it's not a one-liner, but use a multi-line block.  That way you can also store the split in a local variable (`mkey, mvalue = lv.match.split(...)`) instead of doing it twice.  I think it's OK, just a bit long.

@ohadlevy there isn't an each_with_index?

ohadlevy replied:
@domcleal @orrabin i should read more or shut up, thanks :)

orrabin commented:
@domcleal fixed your comment, hopefully this is a bit more readable now

domcleal commented on app/services/classification/base.rb:
```diff
@@ -39,7 +39,11 @@ def values_hash(options = {})
       class_parameters.each do |key|
         lookup_values_for_key = all_lookup_values.where(:lookup_key_id => key.id, :use_puppet_default => false)
-        sorted_lookup_values = lookup_values_for_key.sort_by { |lv| key.path.index(lv.match.split(LookupKey::EQ_DELM).first) }
+        sorted_lookup_values = lookup_values_for_key.sort_by { |lv|
```
use do/end for multi-line blocks (I wonder if rubocop catches that)

orrabin replied:
rubocop didn't like the end bracket, it wanted it to be under the `sorted_lookup_values`

domcleal commented:
Merged as daf34ae, thanks @orrabin!
